src/test/shared/vscode/window.ts

```diff
@@ -0,0 +1,64 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { SeverityLevel, ShownMessage, TestMessage } from './message'
+
+type Window = typeof vscode.window
+
+export interface TestWindow {
```

Comment:
This testing construct is event-based as opposed to the old 'fake' window which was static. UI is largely an event-driven thing, so I think it's reasonable to model test doubles this way. Test code like this requires more effort to create initially but I think it's worth it for things that are used frequently (e.g. notifications).

Reply:
Just to add a bit more of a brain dump here:

Pros of this design:
* Hides testing implementation from test code. It becomes easy to drop in the real VS Code API if we figure out a way to programmatically manipulate it (preferably without manipulating fields not specified in the public API, though it shouldn't be discounted either). Unit tests could become integration tests by implementing the interface with a full-scale UI testing framework.
* Greater flexibility in the ways to test. An async design allows for testing full flows, so it works well for small and large tests.
* Tests read more like a user interaction, e.g. dialog opens -> they read the text -> they select an option
* We are 1:1 with the VS Code API, so this impl. could replace the real `vscode.window` at test time, removing the need for DI (and thus all the `window` params)

Cons:
* Requires implementing test doubles/fakes. More code is not a good thing.
* More annoying to debug when a test case fails. It's easy to get into a 'hung' state with event-driven tests. This can be mitigated with Mocha hooks (but that's more logic!).
* Event-based testing is not as ergonomic for some use-cases as the flow is non-linear. We should support a declarative-style in addition to events, though again that's more code.

Reply:
> More annoying to debug when a test case fails. It's easy to get into a 'hung' state with event-driven tests. This can be mitigated with Mocha hooks (but that's more logic!).

This tends to be a real killer.

Tests are already running vscode; if we can find a way to control the dialogs from test code then maybe we don't need the fakes.